snuba/settings/__init__.py

```diff
@@ -155,11 +156,14 @@ def _load_settings(obj: MutableMapping[str, Any] = locals()) -> None:
             module_format = (
                 ".%s" if settings.startswith("settings_") else ".settings_%s"
             )
-            settings_module = importlib.import_module(module_format % settings, "snuba")
+            settings_module = importlib.import_module(
+                module_format % settings, "snuba.settings"
+            )
 
         for attr in dir(settings_module):
             if attr.isupper():
                 obj[attr] = getattr(settings_module, attr)
 
 
 _load_settings()
+_validate_settings(locals())
```

Comment:
I think you are going to run into the same problem I pointed out here:
https://github.com/getsentry/snuba/pull/1820#discussion_r618762267

Making the settings file depend on the validaiton logic means the settings file will depend on all data structure you need for validaiton, which can be application data structures (see the list of topics or storage keys). Now this will imply that none of the application data structures you may need to use to validate the config can be defined in a module that depends on settings. 
We can move towards that direction (having dedicated modules that define data types or dedicated modules that define the structure of the config elements) but we are not there and you are going to run into circular dependencies when adding the validation logic.

In order to go this way we need to first separate all the data structures that can be referenced in the config from the logic that uses them.

Reply:
I think having settings depend on data structures used for validation is fine. Since we have no validation yet, we can easily make sure that there are no circular dependencies as we slowly add the validation here. I can't imagine there being many dependencies here apart from the simple topic, storage enums.

Alternatively we can call the validation from outside settings when any process starts up, but I don't want to go down this route since there is no shared entrypoint shared between api/consumers/replacers/etc, and we might need to call this all over the place.

Reply:
> I think having settings depend on data structures used for validation is fine. 

I am not arguing against this. The problem is that, in the first use case for config validation (the Topics enum), we are already loading a lot of stuff that could need settings even if today it happens not to.
https://github.com/getsentry/snuba/blob/master/snuba/utils/streams/__init__.py#L1-L7
as streams/topic.py is inside streams
https://github.com/getsentry/snuba/pull/1820/files#diff-fee8bd5753b6822df5d1569a8920c171f77c26f7ea7f087966666a0c50834de8R5

The safe way to ensure we allow the settings validation logic not to depend on something that may need settings is to move those type definitions in a submodule of settings. Or put them anywhere that they are separated enough from the business logic that we can be confident no dependency on settings will be there.

Reply:
This particular PR doesn't do any of this. In the other PR, the Topics enum is in a file by itself now. I can move it to a different location if you like (suggestions welcome), it's current location in streams just seemed the most logical home to me. Putting it inside settings just doesn't seem right to me.

One other option (which I personally quite like, but suspect others won't) could be to just duplicate the list of valid topics that can be configured by the user within settings.

Reply:
Agreed offline that we won't make the validators depend on the application logic.